### DISCUSSION TITLE
How to make Io call C function

### DISCUSSION THREAD

**Discussion by iacore** [stevedekorte/io#472] [Q&A] [answered]
How do I complete the following code?

```
#include <stdio.h>
#include "IoState.h"

int main(int argc, const char *argv[]) {
    IoState *self = IoState_new();

    IoObjectData data;
    IoState_init(self);

    /* What should I do here? */

    IoState_doCString_(self, "hello");
    IoState_free(self);
    return 0;
}

IoObject* hello(IoObject *self, IoObject *locals, IoMessage *m) {
    puts("Hello world!");
    return self;
}
```

**Comment by stevedekorte** [comment]
Hey, Io's binding API is designed around objects so it's best to create an object in which to place you're C functions. The object can also be used to store any C state/context pointers associated with the C library you are binding. Check out the Date binding for an example:  

https://github.com/IoLanguage/io/blob/master/libs/iovm/source/IoDate.h
https://github.com/IoLanguage/io/blob/master/libs/iovm/source/IoDate.c

If the C library you are binding to involves allocating C structs or other pointers that need to be passed around, it's best to create an Io wrapper class for each. The ODE addon is an example of that: 

https://github.com/IoLanguage/ODE/tree/master/source

It has an Io object wrapper for each of the object types in the ODE library.

**Comment by stevedekorte** [comment, marked as answer]
It's been a while since I've used the C binding so I asked ChatGPT4 for a simple example with just one C function added to the lobby. Here's what it wrote (which looks reasonable, but I'm not sure it's correct). This just adds a method for the function and calls it using doCString so it's not and general purpose binding:

#include <stdio.h>
#include "IoState.h"
#include "IoMessage.h"
#include "IoNumber.h"
#include "IoObject.h"

IoObject* hello(IoObject *self, IoObject *locals, IoMessage *m) {
    puts("Hello world!");
    return self;
}

int main(int argc, const char *argv[]) {
    IoState *self = IoState_new();

    IoObject *lobby = IoState_lobby(self);
    IoState_init(self);

    IoMethodTable methodTable[] = {{"hello", hello}, {NULL, NULL}};

    IoObject_addMethodTable_(lobby, methodTable);

    IoState_doCString_(self, "hello();");
    IoState_free(self);
    return 0;
}

**Reply by iacore** [reply]
wow it's correct! What have you fed to ChatGPT4?

**Comment by iacore** [comment]
Here's the full anwser

```
#include <stdio.h>
#include "IoObject.h"


IoObject* hello(IoObject *self, IoObject *locals, IoMessage *m) {
    puts("Hello world!");
    return self;
}

int main(int argc, const char *argv[]) {
    IoState *self = IoState_new();

    IoObject *lobby = IoState_lobby(self);

    IoObject_addMethod_(lobby, IoState_symbolWithCString_(self, "hello"), hello);

    // this also works
    // IoMethodTable methodTable[] = {{"hello", hello}, {NULL, NULL}};
    // IoObject_addMethodTable_(lobby, methodTable);

    IoState_doCString_(self, "hello print");
    IoState_free(self);
    return 0;
}
```

**Comment by stevedekorte** [comment]
Thanks. Glad it worked. It's got me thinking I should have GPT4 write the unfinished binding section of the Io Programming Guide. Btw, here's the session info: https://chat.openai.com/share/f4dbb24b-7c93-4105-ac55-0b3599622c29

**Reply by iacore** [reply]
It seems like GPT-4 has read and remembers the source code of io.

**Comment by ales-tsurko** [comment]
There is also a commented example of writing native addon [here](https://github.com/IoLanguage/NullAddon), which might be helpful

**Reply by iacore** [reply]
I can't find `AddonBulider`
